### PR TITLE
Async now hides `new` and provides safe `create` factory method

### PR DESCRIPTION
**Pull request by jdantonio**

### CODE REVIEW

File: lib/concurrent/async.rb [hunk 1]
```diff
@@ -42,12 +125,35 @@ def validate_argc(obj, method, *args)
...
+      base.extend(ClassMethods)
+      base.send(:private_class_method, :new)
+      super(base)
+    end
+
+    # @!visibility private
+    module ClassMethods
+      def create(*args, &block)
+        obj = self.send(:new, *args, &block)
+        obj.send(:init_synchronization)
```
Comment (pitr-ch) [inline lib/concurrent/async.rb]: if user calls one of the async methods in initialize (even that it's a bad practice) this ca fail. Swapping the order: `allocate`, `init_synchronization`, `initialize` should fix it.

Comment (pitr-ch) [inline lib/concurrent/async.rb]: Hiding the new method breaks API, the `create` method could just override the original `new`, without the name change.

Comment (jdantonio) [inline lib/concurrent/async.rb]: I know it breaks the API, but I think that overriding `new` is a questionable practice. Using a factory method makes it clear that the object creation process has been modified. I'll add some deprecation warnings to and backward-compatable methods.

File: lib/concurrent/async.rb [hunk 2]
```diff
@@ -42,12 +125,35 @@ def validate_argc(obj, method, *args)
...
+    def self.extended(base)
+      base.extend(ClassMethods)
+      base.send(:private_class_method, :new)
+      super(base)
+    end
+
+    # @!visibility private
+    module ClassMethods
+      def create(*args, &block)
+        obj = self.send(:new, *args, &block)
```
Comment (pitr-ch) [inline lib/concurrent/async.rb]: btw just `new *args, &block` should work here.

**Pull request merged**